### DISCUSSION TITLE
How to avoid over fitting in Tabular Prediction using Autogluon

### DISCUSSION THREAD

**Discussion by jimmychordia** [autogluon/autogluon#1273] [Q&A] [answered]
When we are creating models using Tabular Predictor, in the training data the model performance is very high (AUC > 0.98) for all all the models created. Please suggest some technique or parameter option to reduce overfitting in training sample.

**Comment by Innixma** [comment, marked as answer]
Please provide a code example of how you are identifying that the AUC is > 0.98. It is not a problem that AutoGluon overfits to training data. You should try held-out test data / validation data to validate AutoGluon with.

**Reply by jimmychordia** [reply]
#This is the tabular predictor code
predictor = TabularPredictor(label='MIS_SYS_UNAPPROVED_IND',eval_metric = 'roc_auc',verbosity=4).fit(X_train,
                             ag_args_fit={'num_gpus': 1},presets='best_quality', num_bag_sets=1, num_stack_levels=1,
                             excluded_model_types = ['KNN', 'XT' ,'NN','CAT','FASTAI']) 

Calling the leaderboard for train and test:
leaderboard_train = predictor.leaderboard(X_train, extra_metrics=['roc_auc'],extra_info=True)
leaderboard_train

model	score_test	        roc_auc
0	RandomForestEntr_BAG_L1	  1
1	RandomForestGini_BAG_L1	  1
2	LightGBMLarge_BAG_L1	         1
3	WeightedEnsemble_L2	         0.999532
4	LightGBM_BAG_L1	          0.999475

leaderboard_test = predictor.leaderboard(test_data, extra_metrics=['roc_auc'], extra_info=True)
leaderboard_test
model	score_test	        roc_auc
0	RandomForestEntr_BAG_L1	  0.85
1	RandomForestGini_BAG_L1	  0.83
2	LightGBMLarge_BAG_L1	         0.78
3	WeightedEnsemble_L2	        0.86
4	LightGBM_BAG_L1	                0.78

We can see that in test the auc is around 85% which is good. But we need to present the results to business where both train and test results are shared. Looking into the train results it seems model is overfitting and hence may not be stable in future versions. Thus if there is any option where we can reduce the train AUC's and make it more comparable to Test AUC's is needed.

**Reply by Innixma** [reply]
Training AUC is not a relevant metric because models such as KNN and Random Forest explicitly memorize the training set (This is expected and is not a concern). What you want to compare is validation AUC, `score_val` in leaderboard.

AutoGluon's multi-layer stacking (best_quality) implements out-of-fold predictions for calculating validation scores. This means you cannot get the validation score by passing the training data to predictor.leaderboard.

Instead, do the following:

```
leaderboard = predictor.leaderboard()
# leaderboard score_val:
# leaderboard['score_val']
```

For more information about out-of-fold predictions and how they work, refer to the AutoGluon-Tabular paper: https://arxiv.org/pdf/2003.06505.pdf

**Reply by jimmychordia** [reply]
Thanks a lot
